src/DependencyInjection/BaldinofRoadRunnerExtension.php

```diff
@@ -157,4 +163,35 @@ private function aliasIfNotExists(ContainerBuilder $container, string $alias, st
 
         $container->setAlias($alias, $id);
     }
+
+    private function registerRpcRelay(ContainerBuilder $container): void
+    {
+        $rpcDsn = parse_url((string) getenv('RR_RPC'));
```

Comment:
We can not do this at the dependency injection container build time. 

The `RR_RPC` variable exists only when roadrunner run the worker, so if we want to warm up the cache before running RoadRunner it won't work (eg: in a Dockerfile).

The DSN parsing could be done in the factory, with the `Metrics` and `RPC` instanciation.

This way, the only added service to the DIC will be `MetricFactory` and `MetricsInterface`, `baldinof_road_runner.rr.rpc_relay` and `Spiral\Goridge\RPC` could be removed.

The factory can also rely on symfony to get the environment variable by injecting something like `'%env(default::RR_RPC)%'`. By using `default::` symfony will pass `null` if the variable does not exist, so we can throw an approprotiate exception if metrics are enabled in the bundle configuration but RPC is disabled in the RoadRunner configuration.

Hope it makes sense for you!

Reply:
Thanks for feedback, fixed.